solidity/contracts/bank/Bank.sol

```diff
@@ -0,0 +1,291 @@
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/// @title Bitcoin Bank
+/// @notice Bank is a central component tracking Bitcoin balances. Balances can
+///         be transferred between holders and holders can approve their
+///         balances to be spent by others. Balances in the Bank are updated for
+///         depositors who deposit their Bitcoin into the Bridge and only the
+///         Bridge can increase balances.
+/// @dev Bank is a governable contract and the Governance can upgrade the Bridge
+///      address.
+contract Bank is Ownable {
+    address public bridge;
+
+    /// @notice The balance of a given account in the Bank. Zero by default.
+    mapping(address => uint256) public balanceOf;
+
+    /// @notice The remaining amount of balance a spender will be
+    ///         allowed to transfer on behalf of an owner using
+    ///         `transferBalanceFrom`. Zero by default.
+    mapping(address => mapping(address => uint256)) public allowance;
+
+    /// @notice Returns the current nonce for EIP2612 permission for the
+    ///         provided balance owner for a replay protection. Used to
+    ///         construct EIP2612 signature provided to `permit` function.
+    mapping(address => uint256) public nonce;
+
+    uint256 public immutable cachedChainId;
+    bytes32 public immutable cachedDomainSeparator;
+
+    /// @notice Returns EIP2612 Permit message hash. Used to construct EIP2612
+    ///         signature provided to `permit` function.
+    bytes32 public constant PERMIT_TYPEHASH =
+        keccak256(
+            "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        );
+
+    event BalanceTransferred(
+        address indexed from,
+        address indexed to,
+        uint256 amount
+    );
+
+    event BalanceApproved(
+        address indexed owner,
+        address indexed spender,
+        uint256 amount
+    );
+
+    event BalanceIncreased(address indexed owner, uint256 amount);
+
+    event BalanceDecreased(address indexed owner, uint256 amount);
+
+    modifier onlyBridge() {
+        require(msg.sender == address(bridge), "Caller is not the bridge");
+        _;
+    }
+
+    constructor() {
+        cachedChainId = block.chainid;
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    /// @notice Allows the Governance to upgrade the Bridge address.
+    /// @dev The function does not implement any governance delay and does not
+    ///      check the status of the Bridge. The Governance implementation needs
+    ///      to ensure all requirements for the upgrade are satisfied before
```

Comment:
In particular, you mean this owner should probably be a `Timelock`, right?

Reply:
Yes, plus I imagine there has to be a functionality pausing reveals in the Bridge and giving the wallet a chance to sweep deposits before the upgrade is finalized. That's why I left it unspecified in the doc, just "ensure all requirements are satisfied".